stdlib/polylinesimp/rdp/rdp.go

```diff
@@ -0,0 +1,227 @@
+	stack = append(stack, []int64{startIndex, endIndex})
+	for len(stack) > 0 {
+		n := len(stack) - 1
+		lineOfInterest := stack[n]
+		stack = stack[:n]
+		startIndex := lineOfInterest[0]
+		endIndex := lineOfInterest[1]
+		maxDistance := 0.0
+		index := startIndex
+		line := ConnectLine{Starting: ipdata[startIndex], Ending: ipdata[endIndex]}
+		for i := index + 1; i < endIndex; i++ {
+
+			distance := line.perpendicularDistanceFromPointToLine(ipdata[i])
+			if distance > maxDistance {
+				index = i
+				maxDistance = distance
+			}
+
+		}
+		if maxDistance > 0.0 {
+			weights[index] = maxDistance
+			stack = append(stack, []int64{startIndex, index})
+			stack = append(stack, []int64{index, endIndex})
+		}
+	}
+	weights[globalStartIndex] = math.Inf(1)
+	weights[globalEndIndex] = math.Inf(1)
+	return weights
+}
+
+//Learning the value of epsilon (maxTolerance) automatically if both the epsilon value and the retention rate have not been given
+
+func DownSampleIpdataAuto(ipdata []Point) []Point {
+	if len(ipdata) <= 2 {
+		return ipdata
+	}
+	line := ConnectLine{Starting: ipdata[0], Ending: ipdata[len(ipdata)-1]}
+
+	farthestPointIndex, farthestPointDistanceFromLine := FindFarthestPoint(line, ipdata)
+	x1 := ipdata[0].xCoordinate
+	x2 := ipdata[len(ipdata)-1].xCoordinate
+	y1 := ipdata[0].yCoordinate
+	y2 := ipdata[len(ipdata)-1].yCoordinate
+	lengthOfLine := math.Sqrt(math.Pow((x2-x1), 2) + math.Pow((y2-y1), 2))
+	slope := 0.0
+	if (x2 - x1) != 0 {
+		slope = (y2 - y1) / (x2 - x1)
+	}
+	phi := math.Atan(slope) //Phi is the angle between the digitized line and the continuous 2D line made with the X-axis
+	tmax := (1 / lengthOfLine) * (math.Abs(math.Sin(phi)) + math.Abs(math.Cos(phi)))
+	dophimaxone := math.Atan((1 / lengthOfLine) * (math.Abs(math.Sin(phi) + math.Cos(phi))) * (1 - tmax + math.Pow(tmax, 2)))
+	dophimaxtwo := math.Atan((1 / lengthOfLine) * (math.Abs(math.Sin(phi) - math.Cos(phi))) * (1 - tmax + math.Pow(tmax, 2)))
+	dophimax := math.Max(dophimaxone, dophimaxtwo) //Maximum deviation
+	window_threshold := lengthOfLine * dophimax
+	if farthestPointDistanceFromLine >= window_threshold {
+		leftSplit := DownSampleIpdataAuto(ipdata[:farthestPointIndex+1])
+		rightSplit := DownSampleIpdataAuto(ipdata[farthestPointIndex:])
+		return append(leftSplit[:len(leftSplit)-1], rightSplit...)
+	}
+
+	return []Point{ipdata[0], ipdata[len(ipdata)-1]}
+
+}
+
+// Given the array of 2D points split them into separate x and y arrays
+
+func (r *Rdp) SplitCoordinates(sampledPoints []Point) (*array.Int, *array.Float) {
+	newts := array.NewIntBuilder(r.alloc)
+	newvs := array.NewFloatBuilder(r.alloc)
+	for i := range sampledPoints {
+		newts.Append(int64(sampledPoints[i].xCoordinate))
+		newvs.Append(sampledPoints[i].yCoordinate)
+	}
+	return newts.NewIntArray(), newvs.NewFloatArray()
+}
+
+// Main implementation
+
+func (r *Rdp) Do(vs *array.Float, ts *array.Float) (*array.Int, *array.Float, error) {
+	r.vs = vs
+	r.ts = ts
+	points := r.ToPoints()
```

Comment:
This logic invalidates the gains had by using the arrow array values. Is it possible to write the algorithm in terms of `r.vs` and `r.ts` using a shared index? That way we do not have to allocate the data multiple times?